### ISSUE TITLE
DataGrid - question on several features

### ISSUE THREAD

**Issue by davidp918** [syncfusion/flutter-widgets#273]
Heyo, I've got some ideas for my datagrid but I have no idea how to achieve them, please help me out, thanks.

1. How can I fix the sorting when I'm also using the datapager with it. It only sorts the rows of the current page instead of all the data.
2. How can I perform sort on initialization?
3. How can I center it, when the columns is not using up all of the width of the device, the table just aligns at the left, I want it to be at the center. (Adding center as the parent widget does not work btw, the datagrid widget seems to just take up the available width and place itself at the left)
4. Finally, I've encountered crashes(due to stack overflow according to the terminal) when I'm running my app on a 64 bit android emulator, while x86 won't. Have you got any cases like that? I'm not sure if that's because of the widget or my code.

**Comment by RenugaSadagoban98** [comment]
Hi @davidp918 , 
Greetings from syncfusion. 

Please find the following details, 

**1. How can I fix the sorting when I'm also using the datapager with it. It only sorts the rows of the current page instead of all the data.** 

Currently, SfDataGrid does not have support for sort whole data when paging is applied. We have already considered this request as a feature. This feature will be implemented in our any of our upcoming release. 
  
Please refer to the below feedback link for further follow up, 
**Feedback link:**  https://www.syncfusion.com/feedback/21023/support-to-sort-whole-data-instead-of-current-page-data 
 we appreciate your patience until then.  

**2. How can I perform sort on initialization?** 

You can refer to the following UG document for perform sort on initialization, 
**UG link:** https://help.syncfusion.com/flutter/datagrid/sorting#programmatic-sorting 

**3. How can I center it, when the columns is not using up all of the width of the device, the table just aligns at the left, I want it to be at the center.** 

You can achieve your requirement, by wrap, the DataGrid inside the container, set the width to the container, and set the alignment as the center. 
Please refer to the following code snippet, 
``
@override 
  Widget build(BuildContext context) { 
    return Scaffold( 
      appBar: AppBar( 
        title: const Text('Syncfusion Flutter DataGrid'), 
      ), 
      body: Container( 
        alignment:Alignment.center , 
        width: 270, 
        child: SfDataGrid( 
          allowSorting: true, 
          source: employeeDataSource, 
          columns: _columns, 
        ), 
      ), 
    ); 
  } 

``
**4. Encountered crashes** 

From the provided information, we have checked your reported scenario, but we are not able to reproduce your reported issue. We have tested a simple sample with the following devices it's working fine.  
 
Device: Redmi 9 prime   
 
**Please provide the following details,**  
1.	Syncfusion updated version  
2.	Issue reproducing steps or video (If possible).
Please check the sample and let us know if you still facing the same issue.
 
**Sample Link:**  
https://www.syncfusion.com/downloads/support/directtrac/general/ze/main-884396379

**Comment by davidp918** [comment]
Hi,

I've got here a sample, though it's a bit long...
```
import 'dart:io';
import 'package:flutter_svg/flutter_svg.dart';
import 'package:http/http.dart' as http;
import 'package:async_builder/async_builder.dart';
import 'package:flutter/material.dart';
import 'dart:math';
import 'package:syncfusion_flutter_core/theme.dart';
import 'package:syncfusion_flutter_datagrid/datagrid.dart';
import 'package:collection/collection.dart';
import 'dart:convert';

void main() async {
  runApp(MyApp());
}

class MyApp extends StatefulWidget {
  @override
  _MyAppState createState() => _MyAppState();
}

class _MyAppState extends State<MyApp> {
  late List<String> _columns;
  late Future<List<Record>> _future;
  Future<List<Record>> getData() async {
    List<Record> res = [];
    try {
      var response = await http.get(
        Uri.parse(
            'https://kztimerglobal.com/api/v2.0/records/top/recent?steamid64=76561198857828380&tickrate=128&stage=0'),
      );
      response.statusCode == HttpStatus.ok
          ? res = recordFromJson(response.body)
          : print('request failed');
    } catch (exception) {
      throw UnimplementedError();
    }
    return res;
  }

  @override
  void initState() {
    super.initState();
    this._columns = ['Map', 'Time', 'Points', 'TPs', 'Date', 'Server'];
    this._future = getData();
  }

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      home: Scaffold(
        body: AsyncBuilder<List<Record>>(
          future: this._future,
          waiting: (context) => Center(child: CircularProgressIndicator()),
          builder: (context, value) {
            return CustomDataTable(
              columns: this._columns,
              data: value!,
            );
          },
        ),
      ),
    );
  }
}

class CustomDataTable extends StatefulWidget {
  final List<dynamic> data;
  final List<String> columns;
  CustomDataTable({
    Key? key,
    required this.data,
    required this.columns,
  }) : super(key: key);

  @override
  _CustomDataTableState createState() => _CustomDataTableState();
}

class _CustomDataTableState extends State<CustomDataTable> {
  List<Map<String, dynamic>> data = [];

  late List<GridTextColumn> _columns;
  late TableDataSource _tableDataSource;
  late int rowsPerPage;

  @override
  void initState() {
    super.initState();
    this.rowsPerPage = 10;
    this._columns = widget.columns
        .map(
          (String column) => GridTextColumn(
              columnName: column,
              label: Container(
                alignment: Alignment.centerLeft,
                padding: EdgeInsets.symmetric(horizontal: 16.0),
                child: Text(
                  '$column',
                  style: TextStyle(
                    color: Colors.white,
                    fontSize: 16,
                  ),
                ),
              ),
              allowSorting: true),
        )
        .toList();
    for (int i = 0; i < widget.data.length; i++) {
      Map<String, dynamic> cur = widget.data[i].toJson();
      cur['index'] = i + 1;
      this.data.add(cur);
    }
    for (int i = 0; i < this.data.length; i++) this.data[i]['index'] = i + 1;
    print(this.data);
    this._tableDataSource = TableDataSource(
      data: this.data,
      columns: widget.columns,
      context: context,
    );
  }

  @override
  Widget build(BuildContext context) {
    Size size = MediaQuery.of(context).size;
    final int rowsPerPage = min(this.rowsPerPage, this.data.length);
    final double dataPagerHeight = 60.0;
    final double contentRowHeight = 44.0;
    final double headerRowHeight = 49.0;
    final double tableHeight = headerRowHeight + rowsPerPage * contentRowHeight;
    return SfDataGridTheme(
      data: SfDataGridThemeData(
        gridLineStrokeWidth: 0,
        sortIconColor: Colors.white,
      ),
      child: Align(
        alignment: Alignment.center,
        child: Container(
          alignment: Alignment.center,
          child: Column(
            children: [
              SizedBox(
                height: tableHeight,
                width: size.width,
                child: SfDataGrid(
                  rowHeight: contentRowHeight,
                  headerRowHeight: headerRowHeight,
                  allowSorting: true,
                  allowMultiColumnSorting: true,
                  allowTriStateSorting: true,
                  showSortNumbers: true,
                  isScrollbarAlwaysShown: false,
                  columns: this._columns,
                  source: this._tableDataSource,
                  verticalScrollPhysics: NeverScrollableScrollPhysics(),
                ),
              ),
              SizedBox(
                height: dataPagerHeight,
                child: SfDataPagerTheme(
                  data: SfDataPagerThemeData(),
                  child: SfDataPager(
                    visibleItemsCount: rowsPerPage,
                    delegate: _tableDataSource,
                    pageCount: this.data.length / rowsPerPage,
                    direction: Axis.horizontal,
                  ),
                ),
              )
            ],
          ),
        ),
      ),
    );
  }
}

class TableDataSource extends DataGridSource {
  List<DataGridRow> _rows = [];
  late int rowsPerPage;
  late BuildContext context;
  late List<Map<String, dynamic>> paginated;
  late List<Map<String, dynamic>> data;
  late List<String> columns;
  final Map<String, String> identifyAttr = {
    'Player': 'player_name',
    'Count': 'count',
    'Average': 'average',
    'Points': 'points',
    'Rating': 'rating',
    'Finishes': 'finishes',
    'Map': 'map_name',
    'Time': 'time',
    'TPs': 'teleports',
    'Date': 'created_on',
    'Server': 'server_name',
    'Points in total': 'totalPoints',
  };

  TableDataSource({
    required List<Map<String, dynamic>> data,
    required List<String> columns,
    required BuildContext context,
  }) {
    this.context = context;
    this.data = data;
    this.columns = columns;
    this.rowsPerPage = 10;
    this.paginated =
        this.data.getRange(0, min(this.data.length, 9)).toList(growable: false);
    this.buildPaginatedDataGridRows();
  }

  void buildPaginatedDataGridRows() {
    this._rows = this
        .paginated
        .map<DataGridRow>((dynamic each) => DataGridRow(
              cells: [
                for (String column in columns)
                  DataGridCell(
                    columnName: column,
                    value: each,
                  )
              ],
            ))
        .toList();
  }

  @override
  Future<bool> handlePageChange(int oldPageIndex, int newPageIndex) async {
    int startIndex = newPageIndex * rowsPerPage;
    int endIndex = startIndex + rowsPerPage;
    if (startIndex < data.length && endIndex <= data.length) {
      this.paginated =
          this.data.getRange(startIndex, endIndex).toList(growable: false);
      buildPaginatedDataGridRows();
      notifyListeners();
    } else {
      this.paginated = [];
    }
    return true;
  }

  @override
  DataGridRowAdapter buildRow(DataGridRow row) {
    List<String> centeredColumns = [
      '#',
      'TPs',
      'Points',
      'Average',
      'Rating',
      'Finishes',
      'Points in total'
    ];

    return DataGridRowAdapter(
      cells: row.getCells().map<Widget>((each) {
        String name = each.columnName;
        return Container(
          padding: EdgeInsets.symmetric(horizontal: 8.0),
          alignment: centeredColumns.contains(name)
              ? Alignment.center
              : Alignment.centerLeft,
          child: getCell(
            each.columnName,
            each.value,
          ),
        );
      }).toList(),
    );
  }

  @override
  int compare(DataGridRow? a, DataGridRow? b, SortColumnDetails sortColumn) {
    final dynamic value1 = a
        ?.getCells()
        .firstWhereOrNull((element) => element.columnName == sortColumn.name)
        ?.value[identifyAttr[sortColumn.name]];
    final dynamic value2 = b
        ?.getCells()
        .firstWhereOrNull((element) => element.columnName == sortColumn.name)
        ?.value[identifyAttr[sortColumn.name]];

    if (value1 == null || value2 == null) {
      return 0;
    }

    if (value1.compareTo(value2) > 0) {
      return sortColumn.sortDirection == DataGridSortDirection.ascending
          ? 1
          : -1;
    } else if (value1.compareTo(value2) == -1) {
      return sortColumn.sortDirection == DataGridSortDirection.ascending
          ? -1
          : 1;
    } else {
      return 0;
    }
  }

  @override
  List<DataGridRow> get rows => _rows;

  Widget vanillaDataCell(dynamic data) => Text(
        '${data == null ? '<Unknown>' : Characters(data.toString())}',
        style: TextStyle(color: Colors.white),
        overflow: TextOverflow.ellipsis,
      );

  Widget buttonDataCell(BuildContext context, dynamic data, String destination,
          dynamic parameter) =>
      data == null
          ? vanillaDataCell(data)
          : InkWell(
              onTap: () => Navigator.of(context)
                  .pushNamed(destination, arguments: parameter),
              child: Text(
                '${Characters(data.toString())}',
                overflow: TextOverflow.ellipsis,
              ));

  Widget getCell(String column, dynamic data) {
    switch (column) {
      case '#':
        return vanillaDataCell(data?['index']);
      case 'Player':
        return buttonDataCell(context, data?['player_name'], '/player_detail',
            [data?['steamid64'], data?['player_name']]);
      case 'Count':
        return vanillaDataCell(data?['count']);
      case 'Average':
        return vanillaDataCell(data?['average']);
      case 'Points':
        return classifyPoints(data?['points']);
      case 'Rating':
        return vanillaDataCell(data?['rating'].toString().substring(0, 6));
      case 'Finishes':
        return vanillaDataCell(data?['finishes']);
      case 'Map':
        return buttonDataCell(context, data?['map_name'], '/map_detail',
            [data['mapId'], data['map_name']]);
      case 'Time':
        return vanillaDataCell(toMinSec(data?['time']));
      case 'TPs':
        return vanillaDataCell(data?['teleports']);
      case 'Date':
        String? date = data?['created_on'].toString();
        return vanillaDataCell(
            '${date?.substring(0, 11)} ${date?.substring(11, 19)}');
      case 'Server':
        return vanillaDataCell(data?['server_name']);
      case 'Points in total':
        return vanillaDataCell(data?['totalPoints']);
      default:
        return vanillaDataCell('error');
    }
  }
}

String recordToJson(List<Record> data) =>
    json.encode(List<dynamic>.from(data.map((x) => x.toJson())));

List<Record> recordFromJson(String str) =>
    List<Record>.from(json.decode(str).map((x) => Record.fromJson(x)));

class Record {
  Record({
    this.id,
    this.steamid64,
    this.playerName,
    this.steamId,
    this.serverId,
    this.mapId,
    this.stage,
    this.mode,
    this.tickrate,
    this.time,
    this.teleports,
    this.createdOn,
    this.updatedOn,
    this.updatedBy,
    this.recordFilterId,
    this.serverName,
    this.mapName,
    this.points,
    this.replayId,
  });

  int? id;
  String? steamid64;
  String? playerName;
  String? steamId;
  int? serverId;
  int? mapId;
  int? stage;
  String? mode;
  int? tickrate;
  double? time;
  int? teleports;
  DateTime? createdOn;
  DateTime? updatedOn;
  int? updatedBy;
  int? recordFilterId;
  String? serverName;
  String? mapName;
  int? points;
  int? replayId;

  factory Record.fromJson(Map<String, dynamic> json) => Record(
        id: json["id"],
        steamid64: json["steamid64"],
        playerName: json["player_name"],
        steamId: json["steam_id"],
        serverId: json["server_id"],
        mapId: json["map_id"],
        stage: json["stage"],
        mode: json["mode"],
        tickrate: json["tickrate"],
        time: json["time"].toDouble(),
        teleports: json["teleports"],
        createdOn: DateTime.tryParse(json["created_on"] ?? ''),
        updatedOn: DateTime.tryParse(json["updated_on"] ?? ''),
        updatedBy: json["updated_by"],
        recordFilterId: json["record_filter_id"],
        serverName: json["server_name"],
        mapName: json["map_name"],
        points: json["points"],
        replayId: json["replay_id"],
      );
  Map<String, dynamic> toJson() => {
        "id": id,
        "steamid64": steamid64,
        "player_name": playerName,
        "steam_id": steamId,
        "server_id": serverId,
        "map_id": mapId,
        "stage": stage,
        "mode": mode,
        "tickrate": tickrate,
        "time": time,
        "teleports": teleports,
        "created_on": createdOn?.toIso8601String(),
        "updated_on": updatedOn?.toIso8601String(),
        "updated_by": updatedBy,
        "record_filter_id": recordFilterId,
        "server_name": serverName,
        "map_name": mapName,
        "points": points,
        "replay_id": replayId,
      };
}

String? toMinSec(double? sec) {
  if (sec == null) {
    return null;
  }
  String min = (sec / 60).truncate().toString().padLeft(2, '0');
  String seconds = (sec % 60).toStringAsFixed(3).toString().padLeft(6, '0');
  return ('$min:$seconds');
}

Widget classifyPoints(int? points) {
  Widget trophy;
  if (points == 1000) {
    return Row(
      children: [
        SizedBox(width: 8),
        goldSvg(26, 26),
      ],
    );
  } else if (points! >= 900) {
    trophy = silverSvg(15, 15);
  } else if (points >= 750) {
    trophy = bronzeSvg(15, 15);
  } else {
    return Text('$points', style: TextStyle(color: Colors.white));
  }
  return Row(
    children: [
      trophy,
      SizedBox(width: 4),
      Text(
        '$points',
        style: TextStyle(color: Colors.white),
      ),
    ],
  );
}

Widget goldSvg(double height, double width) => SvgPicture.asset(
      'assets/icon/trophy.svg',
      height: height,
      width: width,
    );

Widget silverSvg(double height, double width) => SvgPicture.asset(
      'assets/icon/silver.svg',
      height: height,
      width: width,
    );

Widget bronzeSvg(double height, double width) => SvgPicture.asset(
      'assets/icon/bronze.svg',
      height: height,
      width: width,
    );
```

So when scrolling horizontally, the debug console will print:
```
Build fingerprint: 'google/sdk_gphone_x86_64_arm64/generic_x86_64_arm64:11/RSR1.201211.001.A1/7054069:userdebug/dev-keys'
Revision: '0'
ABI: 'x86_64'
Timestamp: 2021-06-08 14:09:24+0800
pid: 7991, tid: 8032, name: 1.ui  >>> com.dp.kzstats <<<
uid: 10153
signal 11 (SIGSEGV), code 128 (SI_KERNEL), fault addr 0x0
    rax 0000733dc12b9820  rbx a81300007378d6c0  rcx 0000733c2661bc51  rdx f83900015a6a8440
    r8  0000000000000000  r9  00007fffcdcc00b8  r10 00000000001724bc  r11 0000000000000297
    r12 ffff020800000000  r13 0000733c364482a0  r14 0000733dc12b97f0  r15 0000733dc12b9810
    rdi 0000733dc12b9390  rsi 00000000000000c0
    rbp 0000733dc12b9800  rsp 0000733c36448220  rip 0000733c37d788d5
backtrace:
      #00 pc 00000000018388d5  /data/app/~~MS7daek8gPlF9C9zFvCl8g==/com.dp.kzstats-dSP_OqjqwZlci-LXif3FnQ==/lib/x86_64/libflutter.so (BuildId: c3b2cfa7e83d2520b3e054bc0558a1f434347ca7)
```

And the device info is printed out above, and if it's build in x86, not 64, it won't crash.

**Comment by RenugaSadagoban98** [comment]
Hi @davidp918,
Thanks for your update.

Currently, we are analyzing your reported issue. We will validate and update you on the details by tomorrow. We appreciate your patience until then.


Regards,
Renuga Devi.

**Comment by RenugaSadagoban98** [comment]
Hi @davidp918, 

Thanks for your patient.  

We have checked your reported issue. After further analysis, this is the known issue at our end in Flutter 2.2.0 version (stable and beta) and debug mode alone. This issue won’t occur in the release mode and the master channel of the Flutter 2.2.0 version. We have already logged your reported issue as a bug. We will include the fix in our 2021 Volume 2 release which is expected to be rolled out on mid of June 2021.  

Please refer to the below feedback link for further follow up,  
**Feedback link:**  https://www.syncfusion.com/feedback/25710/application-is-closed-when-scroll-the-datagrid-in-debug-mode-and-flutter-2-2 
 
We thank you for your support and appreciate your patience in waiting for this update. Please get in touch with us if you would require any further assistance.  


Regards,
Renuga devi S

**Comment by fathah** [comment]
Please add a similar feature as in Flutter built-in ```PaginatedDataTable``` with Checkbox column at the beginning and select-all checkbox in the column header.

<a href="https://ibb.co/L6TK607"><img src="https://i.ibb.co/bL4tL69/paginated.png" alt="paginated" border="0"></a>

Thanks

**Comment by RenugaSadagoban98** [comment]
Hi @fathah ,

The reported issue is the framework level issue and it has been fixed in Flutter 2.2.2 version.  We can't able to reproduce the issue. Please upgrade the flutter version and revert us if you are still facing the same issue. 
We will happy to assist you.

Regards, 
Renuga devi.

**Comment by RenugaSadagoban98** [comment]
Hi @davidp918 , 
  
Thanks for your patience. 


We are glad to announce that our Essential Studio 2021 Volume 3 Main Release V19.3.0.43 is rolled out and is available for download under the following link. 
  
https://www.syncfusion.com/forums/169291/essential-studio-2021-volume-3-main-release-v19-3-0-43-is-available-for-download 
 
Refer the following UG links for hole data sorting feature, 
UG link: https://help.syncfusion.com/flutter/datagrid/paging#sort-all-the-rows-instead-of-rows-available-in-a-page

 
  
We thank you for your support and appreciate your patience in waiting for this release. Please get in touch with us if you would require any further assistance. 

Regards, 
Renugadevi
